### ISSUE TITLE
Multi-level Drop down

### ISSUE THREAD

**Issue by ominibyte** [salim-lachdhaf/searchable_dropdown#428]
**Is your feature request related to a problem? Please describe.**
So, I'm not sure if this feature is already supported or if there are other ways of achieving this.

I have a peculiar case where I need to "expand" a list item to show other items of the same type as seen in the screenshot below. The second item in the list holds other addresses that I would want the dropdown to switch to when clicked.

<img width="572" alt="Screen Shot 2022-06-03 at 11 53 22 AM" src="https://user-images.githubusercontent.com/7020327/171902590-748971e6-5ad0-4055-bcef-fc9a939ff6da.png">

Currently, when an Item is clicked, the dropdown dissapears.

**Describe the solution you'd like**
There are possibly a few ways to approach the solution:
One option could be introducing two other parameters to `DropdownSearch` such as:
```
DropdownSearch<T>(
  canExpand: (T item) => false,  // Returns false by default 
  onExpand: (T item) async => [], // Similar setup as asyncItems but with the item as argument
)
```
The `canExpand` function can be called when an item is clicked. If it can expand then the `onExpand` is called to change the list items else the `onChanged` is called to signal item selection change.

**Comment by salim-lachdhaf** [comment]
i see not why you cannot do that with the actual version.

as first idea, you can create a StateFullWidget that will be your `DropdownSearchBuilder`  , that contains expand widget of the same type as your DropdownSearch, by selecting an item of the expanded list you can set it dynamically as selected item and close the popup programmatically ...

check examples page for more details how to use dynamic selection ...

**Comment by salim-lachdhaf** [comment]
using version 5.0.2, use `interceptCallBacks`.
this is an full example:

```
import 'package:dropdown_search/dropdown_search.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'dropdownSearch Demo',
      //enable this line if you want test Dark Mode
      //theme: ThemeData.dark(),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final myKey = GlobalKey<DropdownSearchState<MultiLevelString>>();
  final List<MultiLevelString> myItems = [
    MultiLevelString(level1: "1"),
    MultiLevelString(level1: "2"),
    MultiLevelString(
      level1: "3",
      subLevel: [
        MultiLevelString(level1: "sub3-1"),
        MultiLevelString(level1: "sub3-2"),
      ],
    ),
    MultiLevelString(level1: "4")
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Padding(
        padding: EdgeInsets.all(8),
        child: DropdownSearch<MultiLevelString>(
          key: myKey,
          items: myItems,
          compareFn: (i1, i2) => i1.level1 == i2.level1,
          popupProps: PopupProps.menu(
            showSelectedItems: true,
            interceptCallBacks: true,//important line
            itemBuilder: (ctx, item, isSelected) {
              return ListTile(
                selected: isSelected,
                title: Text(item.level1),
                trailing: item.subLevel.isEmpty
                    ? null
                    : (item.isExpanded
                        ? IconButton(
                            icon: Icon(Icons.arrow_drop_down),
                            onPressed: () {
                              item.isExpanded = !item.isExpanded;
                              myKey.currentState?.updatePopupState();
                            },
                          )
                        : IconButton(
                            icon: Icon(Icons.arrow_right),
                            onPressed: () {
                              item.isExpanded = !item.isExpanded;
                              myKey.currentState?.updatePopupState();
                            },
                          )),
                subtitle: item.subLevel.isNotEmpty && item.isExpanded
                    ? Container(
                        height: item.subLevel.length * 50,
                        child: ListView(
                          children: item.subLevel
                              .map(
                                (e) => ListTile(
                                  selected: myKey.currentState?.getSelectedItem
                                          ?.level1 ==
                                      e.level1,
                                  title: Text(e.level1),
                                  onTap: () {
                                    myKey.currentState?.popupValidate([e]);
                                  },
                                ),
                              )
                              .toList(),
                        ),
                      )
                    : null,
                onTap: () => myKey.currentState?.popupValidate([item]),
              );
            },
          ),
        ),
      ),
    );
  }
}

class MultiLevelString {
  final String level1;
  final List<MultiLevelString> subLevel;
  bool isExpanded;

  MultiLevelString({
    this.level1 = "",
    this.subLevel = const [],
    this.isExpanded = false,
  });

  MultiLevelString copy({
    String? level1,
    List<MultiLevelString>? subLevel,
    bool? isExpanded,
  }) =>
      MultiLevelString(
        level1: level1 ?? this.level1,
        subLevel: subLevel ?? this.subLevel,
        isExpanded: isExpanded ?? this.isExpanded,
      );

  @override
  String toString() => level1;
}
```

**Comment by ominibyte** [comment]
Sorry this is coming late. I only just realized my email reply was not added as a comment:
...

@salim-lachdhaf  Thanks a lot but it looks like this doesn’t solve my problem. I want to replace the entire list items with brand new items from a remote server when the expandable item is clicked.

I was able to get it working using an ugly hack by first rejecting the selection of “expandable” items with the `onBeforeChange` callback, then inside the callback I set a state value in my state full widget with that expandable item. Next, I reopen the popup using the global key. In the `onBeforePopupOpening` I then change the search field controller value when a state value has been set which immediately triggers a call to `asyncItems` where I check if a state value for an expandable item has been set, I use that to query the server as opposed to the search text. Finally, I clear the state value when `onChanged` is called for non expandable items.

The only issue now is Flutter complaining about duplicate Global Keys due to all the dirty tricks. So any help to get this working smoothly using another approach will be very much appreciated.

On another note, to solving this global key issue, I was wondering if you will be willing to add a controller pattern. It will definitely help to solve my issue as the controller will be swapped automatically if the widget tree changes.

**Comment by ominibyte** [comment]
Thanks a lot but it looks like this doesn’t solve my problem. I want to
replace the entire list items with brand new items from a remote server
when the expandable item is clicked.

I was able to get it working using an ugly hack by first rejecting the
selection of “expandable” items with the onBeforeChange callback, then
inside the callback I set a state value in my state full widget with that
expandable item. Next, I reopen the popup using the global key. In the
onBeforePopupOpening I then change the search field controller value when a
state value has been set which immediately triggers a call to asyncItems
where I check if a state value for an expandable item has been set, I use
that to query the server as opposed to the search text. Finally, I clear
the state value when onChanged is called for non expandable items.

The only issue now is Flutter complaining about duplicate Global Keys due
to all the dirty tricks. So any help to get this working smoothly using
another approach will be very much appreciated.

On another note, to solving this global key issue, I was wondering if you
will be willing to add a controller pattern is possible. It will definitely
help to solve my issue as the controller will be swapped automatically if
the widget tree changes.
